From: Philippe Coval <philippe.coval@vates.tech>
Date: Wed, 18 Feb 2026 14:07:42 +0100
Subject: remove-ioapic

Don't create an IOAPIC when running under Xen

Xen emulates the IOAPIC rather than QEMU so don't create an IOAPIC
device in QEMU. The existing device had no effect anyway because reads and
writes to the IOAPIC are intercepted by Xen so the device was never set
up. IRQs generated by devices emulated by QEMU should either call
xendevicemodel_set_isa_irq_level or xendevicemodel_set_pci_intx_level.
---
 hw/i386/pc_piix.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/hw/i386/pc_piix.c b/hw/i386/pc_piix.c
index 1bd70d1..1fa4f5e 100644
--- a/hw/i386/pc_piix.c
+++ b/hw/i386/pc_piix.c
@@ -216,7 +216,7 @@ static void pc_init1(MachineState *machine,
 
     pc_i8259_create(isa_bus, gsi_state->i8259_irq);
 
-    if (pcmc->pci_enabled) {
+    if (pcmc->pci_enabled && !xen_enabled()) {
         ioapic_init_gsi(gsi_state, "i440fx");
     }
 
